controllers/devworkspace/solver/che_routing.go

```diff
@@ -474,12 +476,38 @@ func getCommonService(objs *solvers.RoutingObjects, dwId string) *corev1.Service
 	return nil
 }
 
-func exposeAllEndpoints(cheCluster *chev2.CheCluster, routing *dwo.DevWorkspaceRouting, objs *solvers.RoutingObjects, ingressExpose func(*EndpointInfo), endpointStrategy EndpointStrategy) *corev1.ConfigMap {
+func getServiceForEndpoint(objs *solvers.RoutingObjects, endpoint dwo.Endpoint) *corev1.Service {
+	endpointServiceName := common.EndpointName(endpoint.Name)
+	for _, svc := range objs.Services {
+		if svc.Name == endpointServiceName {
+			return &svc
+		}
+	}
+	return nil
+}
+
+// Returns the appropriate service to use when exposing an endpoint.
+// If the appropriate service could not be found in the given routing objects, an error is returned.
+//
+// Endpoints with the "discoverable" attribute set have their own service that should be used.
+// Endpoints that do not set the "discoverable" attribute should use the common service associated with the workspace.
+func determineEndpointService(objs *solvers.RoutingObjects, endpoint dwo.Endpoint, commonService *corev1.Service) (*corev1.Service, error) {
+	if endpoint.Attributes.GetBoolean(string(dwo.DiscoverableAttribute), nil) {
+		endpointService := getServiceForEndpoint(objs, endpoint)
+		if endpointService == nil {
+			return nil, fmt.Errorf("could not find endpoint-specfic service for endpoint '%s'", endpoint.Name)
+		}
+		return endpointService, nil
+	}
+	return commonService, nil
+}
+
+func exposeAllEndpoints(cheCluster *chev2.CheCluster, routing *dwo.DevWorkspaceRouting, objs *solvers.RoutingObjects, ingressExpose func(*EndpointInfo), endpointStrategy EndpointStrategy) (*corev1.ConfigMap, error) {
 	wsRouteConfig := gateway.CreateEmptyTraefikConfig()
 
 	commonService := getCommonService(objs, routing.Spec.DevWorkspaceId)
 	if commonService == nil {
-		return nil
+		return nil, fmt.Errorf("could not find the common service for workspace '%s'", routing.Spec.DevWorkspaceId)
```

Comment:
Can we order functions in the file like so?
`exposeAllEndpoints`
`determineEndpointService` 
`getEndpointService`

Reply:
I assume you mean just moving the functions around in the `che_routing.go` file right? If so, yes! Though I think it's worth also moving [`getCommonService`](https://github.com/eclipse-che/che-operator/blob/main/controllers/devworkspace/solver/che_routing.go#L467-L475)

Reply:
Yes, thank you

Reply:
Done, though I might have gotten your order wrong (I was confused by `determineEndpointService`). Let me know if you want me to re-order things.

Reply:
@dkwon17 did I correctly address this in my last commit https://github.com/eclipse-che/che-operator/pull/1903/commits/fc373a1b05fb9bf590be2a2136132eb7d94b3b54?

Reply:
Could we update the function names?

`getServiceForEndpoint` and `getEndpointService` sound a bit too similar IMO, how about

`getServiceForEndpoint` to `getEndpointService`
`getEndpointSerive` to `getService`

?

Reply:
> getServiceForEndpoint and getEndpointService sound a bit too similar IMO,

I agree, what do you think about the renames in https://github.com/eclipse-che/che-operator/pull/1903/commits/df6fd02aaa1cfbebd6d22aaa4ac5e68195df5f65?

- `getServiceForEndpoint` -> `getEndpointService`
- `getEndpointService` -> `determineEndpointService` (I think this makes it a bit more clear that we're getting the appropriate service to use for an endpoint)

Otherwise, I'll just change `getEndpointService` -> `getService` as you initially suggested :)

Reply:
> getServiceForEndpoint -> getEndpointService
getEndpointService -> determineEndpointService (I think this makes it a bit more clear that we're getting the appropriate service to use for an endpoint)

Ok, lets do that

Reply:
Sounds good :) if you're +1 for this PR, I'll squash my fixup commit and it'll be good to merge